src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php

```diff
@@ -415,14 +435,31 @@ protected function getAliasedPivotColumns()
         // relationships when they are retrieved and hydrated into the models.
         $columns = [];
 
-        foreach (array_merge($defaults, $this->pivotColumns) as $column) {
+        foreach (array_merge($defaults, $this->getPivotModelColumns(), $this->pivotColumns) as $column) {
             $columns[] = $this->table.'.'.$column.' as pivot_'.$column;
         }
 
         return array_unique($columns);
     }
 
     /**
+     * Retrieve necessary columns from the custom pivot model.
+     *
+     * @return array
+     */
+    protected function getPivotModelColumns()
+    {
+        $columns = [];
+
+        if ($this->using) {
+            $customPivotClass = $this->using;
+            $columns = $customPivotClass::getPivotColumns();
```

Comment:
The reasoning behind this is that using the traditional method, you would need to defing the columns in every model using the custom pivot. So if you are doing a 3-way pivot, you would have to define the columns in up to 3 different models. That's bad duplication in my mind. The new method allows you to define the columns on the pivot model so you only have to define the columns in one sensible location.